Fix handling of skip_dir, skip_file & sync_list config option handling

abraunegg opened the pull request:
* config option skip_dir should only be applied to directories
* config option skip_file should only be applied to files
* If 'directory' is in 'skip_file' then upload skip reason erroneously lists sync_list as reason for skipping

abraunegg commented:
@norbusan 
To validate this, add to 'skip_file' a directory to skip - it will be skipped, but reported that it is skipped as part of 'skip_file' .. which is not technically correct.

norbusan commented:
Hmm, maybe I am missing something, but with the current patch I still don't get files in dir skip etc ignored:
```
$ ./onedrive --version
onedrive v2.3.10-3-g4c84093
$ cat ~/.config/onedrive-test/config 
sync_dir = "/home/norbert/OneDriveTest"
skip_dir = "*test*|FileInSkipDir"
skip_file = "DirInSkipFile"
$ ls ~/OneDriveTest
... DirInSkipFile/ ... FileInSkipDir ...
$ ./onedrive --confdir ~/.config/onedrive-test --synchronize -v
...
Uploading new items of .
Skipping item - excluded by skip_dir config: ./Test
Skipping item - path excluded by user config: ./FileInSkipDir
Skipping item - path excluded by user config: ./DirInSkipFile
...
```

This isn't as I expected to be?

abraunegg commented:
With patch - thats as expected - without patch, dir is skipped as file.

With patch, as dir entered in skip_file - the skip is not processed in skil file checks thus falls to a default - its been skipped - but by what ..

abraunegg commented:
**onedrive folder**
```
[alex@centos7full OneDrivePersonal]$ ls -la
total 8
drwxrwxr-x.  3 alex alex   48 Oct  8 13:25 .
drwx------. 31 alex alex 4096 Oct  8 13:25 ..
drwxrwxr-x.  2 alex alex    6 Oct  8 13:25 dirinskipfile
-rw-rw-r--.  1 alex alex    6 Oct  8 13:25 fileinskipdir
[alex@centos7full OneDrivePersonal]$ 
```

**config**
```
onedrive version                    = v2.3.10-2-g8b969de
Config path                         = /home/alex/.config/onedrive-personal/
Config file found in config path    = true
Config option 'check_nosync'        = false
Config option 'sync_dir'            = /home/alex/OneDrivePersonal
Config option 'skip_dir'            = fileinskipdir
Config option 'skip_file'           = ~*|.~*|*.tmp|dirinskipfile
Config option 'skip_dotfiles'       = false
Config option 'skip_symlinks'       = false
Config option 'monitor_interval'    = 45
Config option 'min_notify_changes'  = 5
Config option 'log_dir'             = /var/log/onedrive/
Config option 'sync_root_files'     = false
Selective sync configured           = false
```

**without patch**
```
Using Config Dir: /home/alex/.config/onedrive-personal/
Deleting the saved status ...
Initializing the OneDrive API ...
Opening the item database ...
All operations will be performed in: /home/alex/OneDrivePersonal
Application version: v2.3.10-2-g8b969de
Account Type: personal
Default Drive ID: 66d53be8a5056eca
Default Root ID: 66D53BE8A5056ECA!101
Remaining Free Space: 5368709120
Fetching details for OneDrive Root
OneDrive Root does not exist in the database. We need to add it.
Added OneDrive Root to the local database
Initializing the Synchronization Engine ...
Syncing changes from OneDrive ...
Applying changes of Path ID: 66D53BE8A5056ECA!101
Uploading differences of .
Processing root
The directory has not changed
Uploading new items of .
Skipping item - path excluded by sync_list: ./dirinskipfile   <-- not right
Skipping item - path excluded by sync_list: ./fileinskipdir   <-- not right
Applying changes of Path ID: 66D53BE8A5056ECA!101
```
**with patch**
```
Using Config Dir: /home/alex/.config/onedrive-personal/
Deleting the saved status ...
Initializing the OneDrive API ...
Opening the item database ...
All operations will be performed in: /home/alex/OneDrivePersonal
Application version: v2.3.10-3-g4c84093
Account Type: personal
Default Drive ID: 66d53be8a5056eca
Default Root ID: 66D53BE8A5056ECA!101
Remaining Free Space: 5368709120
Fetching details for OneDrive Root
OneDrive Root does not exist in the database. We need to add it.
Added OneDrive Root to the local database
Initializing the Synchronization Engine ...
Syncing changes from OneDrive ...
Applying changes of Path ID: 66D53BE8A5056ECA!101
Uploading differences of .
Processing root
The directory has not changed
Uploading new items of .
Skipping item - path excluded by user config: ./dirinskipfile   <-- correct because dir is in skip_file
Skipping item - path excluded by user config: ./fileinskipdir   <-- correct because file is in skip_dir
Applying changes of Path ID: 66D53BE8A5056ECA!101
```

**without patch when syncing folder from OneDrive called dirinskipfile**
```
Using Config Dir: /home/alex/.config/onedrive-personal/
Deleting the saved status ...
Initializing the OneDrive API ...
Opening the item database ...
All operations will be performed in: /home/alex/OneDrivePersonal
Application version: v2.3.10-2-g8b969de
Account Type: personal
Default Drive ID: 66d53be8a5056eca
Default Root ID: 66D53BE8A5056ECA!101
Remaining Free Space: 5368709120
Fetching details for OneDrive Root
OneDrive Root does not exist in the database. We need to add it.
Added OneDrive Root to the local database
Initializing the Synchronization Engine ...
Syncing changes from OneDrive ...
Applying changes of Path ID: 66D53BE8A5056ECA!101
Skipping item - excluded by skip_file config: dirinskipfile   <-- not right - dir should not be skipped by skip_file
Uploading differences of .
Processing root
The directory has not changed
Uploading new items of .
Skipping item - path excluded by sync_list: ./dirinskipfile   <-- not right
Skipping item - path excluded by sync_list: ./fileinskipdir   <-- not right
Applying changes of Path ID: 66D53BE8A5056ECA!101
```

**with patch when syncing folder from OneDrive called dirinskipfile but in skip_file**
```
Using Config Dir: /home/alex/.config/onedrive-personal/
Deleting the saved status ...
Initializing the OneDrive API ...
Opening the item database ...
All operations will be performed in: /home/alex/OneDrivePersonal
Application version: v2.3.10-3-g4c84093
Account Type: personal
Default Drive ID: 66d53be8a5056eca
Default Root ID: 66D53BE8A5056ECA!101
Remaining Free Space: 5368709120
Fetching details for OneDrive Root
OneDrive Root does not exist in the database. We need to add it.
Added OneDrive Root to the local database
Initializing the Synchronization Engine ...
Syncing changes from OneDrive ...
Applying changes of Path ID: 66D53BE8A5056ECA!101
Skipping item - path excluded by user config: dirinskipfile    <-- correct, but this is because dir is in skip_file
Uploading differences of .
Processing root
The directory has not changed
Uploading new items of .
Skipping item - path excluded by user config: ./dirinskipfile    <-- correct, but this is because dir is in skip_file
Skipping item - path excluded by user config: ./fileinskipdir    <-- correct, but this is because file is in skip_dir
Applying changes of Path ID: 66D53BE8A5056ECA!101
```

then switch config around (dir in skip_dir, file in skip_file)

```
Using Config Dir: /home/alex/.config/onedrive-personal/
Deleting the saved status ...
Initializing the OneDrive API ...
Opening the item database ...
All operations will be performed in: /home/alex/OneDrivePersonal
Application version: v2.3.10-3-g4c84093
Account Type: personal
Default Drive ID: 66d53be8a5056eca
Default Root ID: 66D53BE8A5056ECA!101
Remaining Free Space: 5368709114
Fetching details for OneDrive Root
OneDrive Root does not exist in the database. We need to add it.
Added OneDrive Root to the local database
Initializing the Synchronization Engine ...
Syncing changes from OneDrive ...
Applying changes of Path ID: 66D53BE8A5056ECA!101
Skipping item - excluded by skip_dir config: dirinskipfile   <-- correct
Skipping item - excluded by skip_file config: fileinskipdir   <-- correct
Uploading differences of .
Processing root
The directory has not changed
Uploading new items of .
Skipping item - excluded by skip_dir config: ./dirinskipfile   <-- correct
Skipping item - excluded by skip_file config: ./fileinskipdir   <-- correct
Applying changes of Path ID: 66D53BE8A5056ECA!101
```

norbusan commented:
Hmmm, I don't understand, sorry. I thought that a file `A` in `skip_dir` will not be skipped? That is what I would expect ...

abraunegg commented:
> Hmmm, I don't understand, sorry. I thought that a file `A` in `skip_dir` will not be skipped? That is what I would expect ...

Correct - it is not, however, there is a 'catch all' check as well going on as well after file | dir checks:
```
selectiveSync.isPathExcluded(path)
```
which does the following:
```
bool isPathExcluded(string path)
{
     return .isPathExcluded(path, paths) || .isPathMatched(path, mask) || .isPathMatched(path, dirmask);
}
```

Which again checks as a 'catch all':
* sync_list
* skip_file
* skip_dir

norbusan commented:
Hmm, but that is not correct, do we agree on this?

abraunegg commented:
Yep .. that the catch all is double checking - not right at all.

abraunegg commented:
@norbusan 
**latest config**
```
onedrive version                    = v2.3.10-3-g4c84093
Config path                         = /home/alex/.config/onedrive-personal/
Config file found in config path    = true
Config option 'check_nosync'        = false
Config option 'sync_dir'            = /home/alex/OneDrivePersonal
Config option 'skip_dir'            = dirinskipdir
Config option 'skip_file'           = ~*|.~*|*.tmp|fileinskipfile
Config option 'skip_dotfiles'       = false
Config option 'skip_symlinks'       = false
Config option 'monitor_interval'    = 45
Config option 'min_notify_changes'  = 5
Config option 'log_dir'             = /var/log/onedrive/
Config option 'sync_root_files'     = false
Selective sync configured           = true
sync_list contents:
syncthis
```
**latest patch**
```
(dmd-2.085.0)[alex@centos7full onedrive-pr686]$ ./onedrive --confdir '~/.config/onedrive-personal/' --synchronize --verbose --resync
Using Config Dir: /home/alex/.config/onedrive-personal/
Deleting the saved status ...
Initializing the OneDrive API ...
Opening the item database ...
All operations will be performed in: /home/alex/OneDrivePersonal
Application version: v2.3.10-3-g4c84093
Account Type: personal
Default Drive ID: 66d53be8a5056eca
Default Root ID: 66D53BE8A5056ECA!101
Remaining Free Space: 5368709120
Fetching details for OneDrive Root
OneDrive Root does not exist in the database. We need to add it.
Added OneDrive Root to the local database
Initializing the Synchronization Engine ...
Syncing changes from OneDrive ...
Applying changes of Path ID: 66D53BE8A5056ECA!101
Skipping item - excluded by skip_dir config: dirinskipdir
Uploading differences of .
Processing root
The directory has not changed
Uploading new items of .
Skipping item - excluded by skip_file config: ./fileinskipfile
Skipping item - excluded by skip_dir config: ./dirinskipdir
Skipping item - excluded by skip_list config: ./skipthis
Applying changes of Path ID: 66D53BE8A5056ECA!101
Skipping item - excluded by skip_dir config: dirinskipdir
```
